[apm/known_issues] Add known issues for ILM to DSL switch (#4192)
elastic/observability-docs: docs/en/observability/apm/known-issues.asciidoc

Link to fix
////

*Upgrading to v8.15.0 may cause APM indices to become unmanaged w.r.t. lifecycle policy* +
_Elastic Stack versions: 8.15.0_ +

> _Fixed in Elastic Stack version 8.15.1_

Author: It does!

// The conditions in which this issue occurs
The issue occurs when Elastic Stack is upgraded to 8.15.0.
The issue will NOT occur on a new cluster creation with version 8.15.0.
The issue occurs because APM Server switched to Elasticsearch APM-data plugin for managing index templates.

The switch also migrated us from Index Lifecycle Management (ILM) to Datastream Lifecycle (DSL) for managing
data retention for APM indices. Due to this migration, any datastream created before 8.15.0 would be without
datastream lifecycle configuration and, when migrated to 8.15.0, would become unmanaged (without any data 
retention configuration). This can be fixed by explicitly configuring APM datastreams with the default

> lifecyle config:

Author: Sounds better, thanks for the suggestions!

[source,txt]
----

Thanks for adding more details, we have identified a bug which is causing this issue. In 8.15.0 we migrated from using ILM to DLM (Datastream Lifecycle) and on update, the existing datastreams are not updated with the DLM settings causing Unmanaged indexes.

I have created a bug report in APM-Server and we are working on a fix. For short-term mitigation, we would need to update all the APM datastream manually with the default data retention:

> 

Author: This is a copy-paste mistake 😮‍💨 , thanks for catching it.

PUT _data_stream/traces-apm-*/_lifecycle
{
  "data_retention": "10d"
}

PUT _data_stream/traces-apm.rum*/_lifecycle
{
  "data_retention": "90d"
}

PUT _data_stream/traces-apm.sampled*/_lifecycle
{
  "data_retention": "1h"
}

PUT _data_stream/metrics-apm.*.1m-*/_lifecycle
{
  "data_retention": "90d"
}

PUT _data_stream/metrics-apm.*.10m-*/_lifecycle
{
  "data_retention": "180d"
}

PUT _data_stream/metrics-apm.*.60m-*/_lifecycle
{
  "data_retention": "390d"
}

PUT _data_stream/metrics-apm.internal-*/_lifecycle
{
  "data_retention": "90d"
}

PUT _data_stream/metrics-apm.app.*/_lifecycle
{
  "data_retention": "90d"
}

PUT _data_stream/logs-apm.*/_lifecycle
{
  "data_retention": "10d"
}
----

> NOTE that while this issue is fixed in 8.15.1, any indices created in 8.15.0 due
> to rollovers would still be unmanaged even after upgrading. The only way to fix
> these is by configuring the datastream lifecycle explicitly as described above.
> Explicitly configuring datastream lifecycle has an added benefit of migrating
> to datastream lifecycle for all APM datastreams.

Reviewer: Is the linked PR the one that fixed the issue?

```suggestion
While this issue is fixed in 8.15.1 (https://github.com/elastic/elasticsearch/pull/112028[elastic/elasticsearch#112028]), any indices created in 8.15.0 due
to rollovers would still be unmanaged even after upgrading. The only way to fix
these is by configuring the data stream lifecycle explicitly as described above.
Explicitly configuring data stream lifecycle has an added benefit of migrating
to data stream lifecycle for all APM data streams.
```

Author: Yes, the link is correct, thanks for adding that.

Link to fix
////

*Upgrading to v8.15.0 may cause APM indices to lose their lifecycle policy* +
_Elastic Stack versions: 8.15.0_ +
_Fixed in Elastic Stack version 8.15.1_

// The conditions in which this issue occurs
The issue only occurs when _upgrading_ the {stack} to 8.15.0.
The issue does _not_ occur when creating a _new_ cluster using 8.15.0.
The issue also does not occur if a custom ILM policy is configured using a custom component template.

// Describe why it happens

> In 8.15.0, APM Server switched to use Data stream Lifecycle to manage data retention

Reviewer: Based on the [Elasticsearch docs that cover this topic](https://www.elastic.co/guide/en/elasticsearch/reference/current/data-stream-lifecycle.html), I think all three words should be lowercase: data stream lifecycle.

```suggestion
In 8.15.0, APM Server switched to use data stream lifecycle to manage data retention
```

for APM indices for new deployments as well as for upgraded deployments with default lifecycle

> configurations. Unfortunately, since any data stream created before 8.15.0 does not have data
> stream lifecycle configuration, such existing data streams become unmanaged for default

Reviewer: ```suggestion
configurations. Unfortunately, since any data stream created before 8.15.0 does not have a data
stream lifecycle configuration, such existing data streams become unmanaged for default
```

lifecycle configurations.

// How to fix it

> Upgrading to 8.15.1 should fix any new indices created for the data stream. However,
> indices created in version 8.15.0 would remain unmanaged if a custom ILM policy is not
> used. One of the following two approaches can be adopted to fix the unmanaged indices:

Reviewer: This feels a bit easier to understand: 

```suggestion
Upgrading to 8.15.1 should fix any new indices created for the data stream. However,
indices created in version 8.15.0 would remain unmanaged if the default ILM policy is
used. One of the following approaches can be adopted to fix the unmanaged indices:
```

1. Manually delete the indices when they are no longer needed.

> 2. Explicitly configure APM data streams with the default data stream lifecycle config,
> note that using this approach would migrate all data streams to data stream lifecycle
> which should be equivalent to the default ILM policies:

Reviewer: ```suggestion
2. Explicitly configure APM data streams with the default data stream lifecycle config.
Using this approach would migrate all data streams to use data stream lifecycles,
which should be equivalent to the default ILM policies:
```

Link to fix
////

*Upgrading to v8.15.0 may cause APM indices to lose their lifecycle policy* +
_Elastic Stack versions: 8.15.0_ +
_Fixed in Elastic Stack version 8.15.1_

// The conditions in which this issue occurs
The issue only occurs when _upgrading_ the {stack} to 8.15.0.
The issue does _not_ occur when creating a _new_ cluster using 8.15.0.
The issue also does not occur if a custom ILM policy is configured using a custom component template.

// Describe why it happens
This happens because APM Server now uses the Elasticsearch apm-data plugin to manage
index templates, and as a result APM Server now uses datastream lifecycle (DSL) instead
of index lifecycle management (ILM) to manage data retention for APM indices.
This means that any datastream created before 8.15.0 does not have DSL configuration
and, when upgraded to 8.15.0, the datastream becomes unmanaged because it has

> no DSL configuration. 

Reviewer 2: Is it important to call out the switch to apm-data plugin? I wonder what the user benefit for this level of detail is, and whether we should simply focus on the switch to DSL; something like:

> in 8.15.0, APM Server switched to use datastream lifecycle (DSL) to manage data retention for APM indices for new deployments as well as for upgraded deployments with default lifecycle configurations. Unfortunately, since any datastream created before `8.15.0` does not have DSL configuration, such existing datastreams become unmanaged.

Author: Sounds good! Updated!

// How to fix it
Upgrading to 8.15.1 should fix any new indices created for the datastream. However,
indices created in version 8.15.0 would remain unmanaged if a custom ILM policy is not

> used. One of the following two approaches can be adopted to fix the unmanaged indices:

Reviewer 2: However, indices created in version 8.15.0 would remain unmanaged **_for default lifecycle configurations_**.
